lib/uri.ex

```diff
@@ -0,0 +1,43 @@
+defmodule Telepoison.URI do
+  @moduledoc """
+  Exposes a function to normalise URIs in a format suitable for usage as Open Telemetry metadata.
+  """
+
+  alias HTTPoison.Request
+
+  @default_route "/"
+
+  @spec infer_route(String.t() | URI.t() | Request.t()) :: String.t()
+  @doc """
+  Infers the route of the provided `uri`, returned in a format suitable for usage as Open Telemetry metadata.
+  """
+  def infer_route(uri) when is_nil(uri) or byte_size(uri) <= 1, do: @default_route
+
+  def infer_route(%URI{path: path}) when is_nil(path) or byte_size(path) <= 1, do: @default_route
+
+  def infer_route(%Request{url: url}) when is_nil(url) or byte_size(url) <= 1, do: @default_route
+
+  def infer_route(uri) when is_binary(uri), do: uri |> URI.parse() |> infer_route()
+
+  def infer_route(%Request{url: url}), do: infer_route(url)
+
+  def infer_route(%URI{path: path}) do
+    segments =
+      path
+      |> String.split("/")
+      |> Enum.filter(&(byte_size(String.trim(&1)) > 0))
+
+    case Enum.count(segments) do
+      1 ->
+        "/#{Enum.take(segments, 1)}"
+
+      count when count > 1 ->
+        segments
+        |> Enum.take(1)
+        |> (&"/#{&1}/:subpath").()
+
+      _ ->
+        @default_route
+    end
+  end
+end
```

Comment:
Is this module and all these function clauses really necessary? It makes it hard to follow and a bit on the "defensive" side IMO.

We call this from the `request` callback, which is called from the basic request implementation https://github.com/edgurgel/httpoison/blob/bdc665199a349981dcda6aeb227587c835597e7a/lib/httpoison/base.ex#L357-L364 with a Request object with a processed url. At that point we'll always have a Request object with an url, not even sure if a `nil` url is possible at that point, but even if it is I would make it a bit clearer with something like this:

```elixir
@spec infer_route_from_request(HTTPoison.Request.t()) :: binary()
def infer_route_from_request(HTTPoison.Request{url: url}) when is_binary(url), do: infer_route_from_url(url)
def infer_route_from_request(HTTPoison.Request{}), do: @default_route

@spec infer_route_from_url(binary()) :: binary()
def infer_route_from_url(url) when is_binary(url) do
    url
    |> URI.parse()
    |> Map.get(:path, "/")
    |> String.split("/", parts: 2, trim: true)
    |> case do
      [] -> @default_route
      [main_path] -> "/" <> main_path
      [main_path, sub_path] -> "/" <> main_path <> "/:subpath"
    end
  end
```

Reply:
Certainly, updated